### DISCUSSION TITLE
Adding existing ot-cli commands to pyspinel for NCP architecture

### DISCUSSION THREAD

**Discussion by rretanubun** [openthread/openthread#8086] [Q&A] [answered]
Hello,

My platform is SiLABS ThunderSense2 and I am trying to use ot-ftd-ncp demo app (thead-1.3)

I am able to communicate with it using pyspinel but a bit surprised that only a very small subset of command is available (`dataset` for example is missing). The plan in mind to to eventually pair the NCP with another (non-linux) embedded device and communicate using spinel.

Is there a good reference commit on how to add ot-cli supported command to pyspinel? Does every ot-cli command added automatically have a SPINEL.PROP defined on the ot-code side that I should match?

The closest recent example I came across is this:

https://github.com/openthread/pyspinel/pull/116/commits/aefaa1e02b8a1603e1fc22aa2c37e82e04dc4cb7

Thank you for everyone's time.

**Comment by jwhui** [comment, marked as answer]
> I am able to communicate with it using pyspinel but a bit surprised that only a very small subset of command is available (dataset for example is missing). The plan in mind to to eventually pair the NCP with another (non-linux) embedded device and communicate using spinel.

`pyspinel` has not been actively supported for some time. As a result, some of the more recent features have not been implemented. The `dataset` command, as you noted, is one of those unimplemented commands.

> Is there a good reference commit on how to add ot-cli supported command to pyspinel? Does every ot-cli command added automatically have a SPINEL.PROP defined on the ot-code side that I should match?

Not every OT CLI command has a matching Spinel property. Note that Spinel was originally designed to be a property-based interface, which is a different model from OT CLI. Spinel does support getting/setting Active Operational Datasets, but how that is done differs from OT CLI.
https://github.com/openthread/openthread/blob/cba1bebf1f1ef3531bfe5397c75c3c6e045218b6/src/lib/spinel/spinel.h#L2684

**Reply by rretanubun** [reply]
Thanks Jonathan, is there a more actively supported alternative for embedded MCUs?
(I did look at wpantund but that approach seems to be more linux-coupled)

**Reply by jwhui** [reply]
I'm not aware of any host-side implementation for NCP on embedded MCUs.

However, host-side implementation for RCP on embedded MCUs does exist. The [ESP Thread Border Router](https://openthread.io/guides/border-router/espressif-esp32) is one example.

**Reply by rretanubun** [reply]
Apologies for not phrasing the question well. If it helps I can open this as a new topic.

I have a working ot-ftd prototype that uses the SoC architecture. 
it uses `DNS-SRP server+client and CoAP` as provided by thread. Interaction happens via ot-cli. 
The next step is to pair this solution to another (non-linux) host-MCU via SPI and let it control it.
The host-MCU is already busy doing other things and RCP is not a good fit for this solution.

So, I am looking for a good starting point to do machine to machine based ot-cli interactions.
Is spinel still the best approach here? 

For example, looking at spinel.h I do not see entries for `SRP Server`. One of the interactions
the solution needs to do is (ot-cli) `srp server host` and get back all the registered end device 
(to get a sense of how many registered devices are present in the network)

If I want to add that to spinel, I am trying to get a sense of `Is that a well-travelled path I can mainline later?` or
`Will this need a lot of non-mainline refactoring?`

Thank you for your time!

**Reply by jwhui** [reply]
Spinel is one approach but, as mentioned, many of the newer features are not implemented including SRP.

An alternative path may be to use the CLI directly.

**Reply by rretanubun** [reply]
Thanks Jonathan. Confirming my understanding from your comment, in the `NCP mode` does the openthread stack has the same set of openthread API and capabilities as the `SOC mode` (going by https://openthread.io/platforms/co-processor#network_co-processor_ncp) - or is it missing the newer features (beyond just adding Spinel access to the APIs)

**Reply by jwhui** [reply]
Exposing newer features via Spinel is a primary issue. However, if you are looking to support a Thread Border Router, then more needs to be done to properly support the Thread Border Router features.

**Comment by simonlingoogle** [comment]
I think it's actually might be a good idea to support CLI commands via NCP for debugging purposes.
But it is controversial and needs notable development efforts (to integrate openthread-cli with openthread-ncp).
